python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,251 @@
+
+from iceberg.types import (
+    BinaryType,
+    BooleanType,
+    DateType,
+    DecimalType,
+    DoubleType,
+    FixedType,
+    FloatType,
+    IntegerType,
+    LongType,
+    StringType,
+    TimestampType,
+    TimestamptzType,
+    TimeType,
+    UUIDType,
+)
+
+HIVE_NULL = "__HIVE_DEFAULT_PARTITION__"
+
+
+@singledispatch
+def from_partition_value_to_py(type_, partition_value_as_str: str):
+    """A generic function function which converts a partition string to a python built-in
+
+    Args:
+        type_(PrimitiveType): An implementation of the PrimitiveType base class
+        partition_value_as_str(str): A string representation of a partition value
+    """
+    if partition_value_as_str is None or partition_value_as_str == HIVE_NULL:
+        return None
+    raise TypeError(f"Cannot convert partition string to python built-in, type {type_} not supported: '{partition_value_as_str}'")
+
+
+@from_partition_value_to_py.register(BooleanType)
+def _(type_, partition_value_as_str: str):
+    if partition_value_as_str is None:
+        return False
+    return partition_value_as_str.lower() == "true"
+
+
+@from_partition_value_to_py.register(IntegerType)
+@from_partition_value_to_py.register(LongType)
+def _(type_, partition_value_as_str: str):
+    return int(float(partition_value_as_str))
+
+
+@from_partition_value_to_py.register(FloatType)
+@from_partition_value_to_py.register(DoubleType)
+def _(type_, partition_value_as_str: str):
+    return float(partition_value_as_str)
+
+
+@from_partition_value_to_py.register(StringType)
+def _(type_, partition_value_as_str: str):
+    return partition_value_as_str
+
+
+@from_partition_value_to_py.register(UUIDType)
+def _(type_, partition_value_as_str: str):
+    return uuid.UUID(partition_value_as_str)
+
+
+@from_partition_value_to_py.register(FixedType)
+def _(type_, partition_value_as_str: str):
+    return bytearray(bytes(partition_value_as_str, "UTF-8"))
+
+
+@from_partition_value_to_py.register(BinaryType)
+def _(type_, partition_value_as_str: str):
+    return bytes(partition_value_as_str, "UTF-8")
+
+
+@from_partition_value_to_py.register(DecimalType)
+def _(type_, partition_value_as_str: str):
+    return Decimal(partition_value_as_str)
+
+
+@singledispatch
+def to_bytes(type_, value: Any) -> bytes:
+    """A generic function which converts a built-in python value to bytes
+
+    Args:
+        type_(PrimitiveType): An implementation of the PrimitiveType base class
+        value(Any): The value to convert to bytes
+    """
+    raise TypeError(f"Cannot serialize value, type {type_} not supported: '{value}'")
+
+
+@to_bytes.register(BooleanType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<?", 1 if value else 0)
+
+
+@to_bytes.register(IntegerType)
+@to_bytes.register(DateType)
+def _(type_, value: Any) -> bytes:
+    return struct.pack("<i", value)
+
+
+@to_bytes.register(LongType)
+@to_bytes.register(TimeType)
+@to_bytes.register(TimestampType)
```

Comment:
Is accepting values as `datetime.time` or `datetime.datetime` something that is handled at a higher level?  Or considered out of scope for the library, or something that wouldn't apply here?